joinmarket-webui/jam
# regtest: setup with local tor in container

This PR is related to https://github.com/joinmarket-webui/joinmarket-webui-docker/pull/47 and can be used for testing it.

**Attention**: After an eventual successful review, and when [joinmarket-webui-docker/#47](https://github.com/joinmarket-webui/joinmarket-webui-docker/pull/47) is merged, the secondary instance should be [changed to extend `master` again](https://github.com/joinmarket-webui/joinmarket-webui/pull/396/files#diff-4ced3dec775ec46f9839466228a4cd4e9d77836be8b066d0f947101400bf45c6L1-R1). Hence, this PR is in "draft" state, till [joinmarket-webui-docker/#47](https://github.com/joinmarket-webui/joinmarket-webui-docker/pull/47) is merged.


In this PR the individual joinmarket containers in the regtest setup get their own local Tor daemon. The external Tor container has been removed.

How to test these changes?
1. For a pure test, remove all other message channels than the directory node onion channel
  - in our setup's case, it's only a single IRC channel -> comment out the section `[MESSAGING:server1]` in both configs([[1]](https://github.com/joinmarket-webui/joinmarket-webui/blob/dev-regtest-with-local-tor/docker/regtest/dockerfile-deps/joinmarket/latest/default.cfg#L91-L99), [[2]](https://github.com/joinmarket-webui/joinmarket-webui/blob/dev-regtest-with-local-tor/docker/regtest/dockerfile-deps/joinmarket/webui-standalone/default.cfg#L91-L99)):
     ```
     #[MESSAGING:server1]
     #host = irc
     #hostid = localhost1
     #channel = joinmarket-pit
     #port = 6667
     #usessl = false
     #socks5 = false
     #socks5_host = localhost
     #socks5_port = 9050
     ```
2. Rebuild the regtest setup with `npm run regtest:rebuild`
3. Start the regtest setup with `npm run regtest:up`
4. Wait for the directory node to come up (see Addendum), and only then init the setup with `npm run regtest:init`
5. Do your manual tests: Running the scheduler (run `npm run regtest:mine` for it to run through) and individual collaborative transactions; watch the logs -> everything should work as expected

##### Addendum
Testing with tor can be very slow sometimes. Occasionally the startup of the directory node can take a couple of minutes (in this example run it took ~1m30s, but most of the time it takes less than 10s):
```
2022-07-12 03:03:58+0000 [JMDaemonServerProtocol,0,127.0.0.1] NoOpProtocolFactory starting on 45295
2022-07-12 03:03:58+0000 [TorControlProtocol,client] 4% Connecting to a relay
[...]
2022-07-12 03:06:32+0000 [TorControlProtocol,client] 90% Done
2022-07-12 03:06:32+0000 [TorControlProtocol,client] OnionLineProtocolFactory starting on 38503
```
It's best to wait for the directory node to have completely started before initializing the setup (see `npm run regtest:logs:directory_node` before executing `npm run regtest:init`). Otherwise the secondary instance might not resend it's offer to the directory node and you'll end up with `ERROR not enough liquidity in the orderbook` when attempting a collaborative transaction. Just make sure that you see a successful connection in the secondary container (-> not `[WARNING]  We failed to connect to directory <direcotrory_node_onion>.onion:5222; trying again.`)

In my local tests, not starting the irc container (while keeping the irc channel config) led to none of the collaborative transactions go through successfully,  although it should have been working theoretically. 
Ended up with a lot of `IRC connection failed` and `Attempting to reconnect...` while never using the p2p onion channel via the directory node. This needs further investigation; it is either a problem with my local setup or with joinmarket itself, but should not be related to Jam.
